generator/altitude_generator.cpp

```diff
@@ -0,0 +1,247 @@
+
+  vector<bool> const & GetAltitudeAvailability() const { return m_altitudeAvailability; }
+
+  TAltitude GetMinAltitude() const { return m_minAltitude; }
+
+  void operator()(FeatureType const & f, uint32_t const & id)
+  {
+    if (id != m_altitudeAvailability.size())
+    {
+      LOG(LERROR, ("There's a gap in feature id order."));
+      return;
+    }
+
+    bool hasAltitude = false;
+    do
+    {
+      if (!routing::IsRoad(feature::TypesHolder(f)))
+        break;
+
+      f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      size_t const pointsCount = f.GetPointsCount();
+      if (pointsCount == 0)
+        break;
+
+      TAltitudes altitudes;
+      bool allPointsValidAltFlag = true;
+      TAltitude minFeatureAltitude = kInvalidAltitude;
+      for (size_t i = 0; i < pointsCount; ++i)
+      {
+        TAltitude const a = m_srtmManager.GetHeight(MercatorBounds::ToLatLon(f.GetPoint(i)));
+        if (a == kInvalidAltitude)
+        {
+          allPointsValidAltFlag = false;
+          break;
+        }
+
+        if (minFeatureAltitude == kInvalidAltitude)
+          minFeatureAltitude = a;
+        else
+          minFeatureAltitude = min(minFeatureAltitude, a);
+
+        altitudes.push_back(a);
+      }
+      if (!allPointsValidAltFlag)
+        break;
+
+      hasAltitude = true;
+      m_featureAltitudes.push_back(make_pair(id, Altitude(move(altitudes))));
+
+      if (m_minAltitude == kInvalidAltitude)
+        m_minAltitude = minFeatureAltitude;
+      else
+        m_minAltitude = min(minFeatureAltitude, m_minAltitude);
+    }
+    while(false);
+
+    m_altitudeAvailability.push_back(hasAltitude);
+  }
+
+  bool HasAltitudeInfo()
+  {
+    return !m_featureAltitudes.empty();
+  }
+
+  void SortFeatureAltitudes()
+  {
+    sort(m_featureAltitudes.begin(), m_featureAltitudes.end(), my::LessBy(&Processor::TFeatureAltitude::first));
+  }
+
+private:
+  generator::SrtmTileManager m_srtmManager;
+  TFeatureAltitudes m_featureAltitudes;
+  vector<bool> m_altitudeAvailability;
+  TAltitude m_minAltitude;
+};
+
+uint32_t GetFileSize(string const & filePath)
+{
+  uint64_t size;
+  if (!my::GetFileSize(filePath, size))
+  {
+    LOG(LWARNING, ("altitudeAvailability", filePath, "size = 0"));
+    return 0;
+  }
+
+  LOG(LINFO, ("altitudeAvailability ", filePath, "size =", size, "bytes"));
+  return size;
+}
+
+void MoveFileToAltitudeSection(string const & filePath, uint32_t fileSize, FileWriter & w)
+{
+  {
+    ReaderSource<FileReader> r = FileReader(filePath);
+    w.Write(&fileSize, sizeof(fileSize));
+    rw::ReadAndWrite(r, w);
+    LOG(LINFO, (filePath, "size is", fileSize));
+  }
+  FileWriter::DeleteFileX(filePath);
+}
+
+void SerializeHeader(TAltitudeSectionVersion version, TAltitude minAltitude,
+                     TAltitudeSectionOffset altitudeInfoOffset, FileWriter & w)
+{
+  w.Write(&version, sizeof(version));
+
+  w.Write(&minAltitude, sizeof(minAltitude));
```

Comment:
Please, create a structure named Header with fields version, minAltitude, altitudeInfoOffset, that can be both serialized and deserialized.